### PR TITLE
fix l1 fee cache, rpc, tracing and tx pool

### PR DESCRIPTION
**Pull request by protolambda**
This PR fixes:
- Use `chainConfig.OptimismConfig != nil` instead of double `hainConfig.OptimismConfig != nil && OptimismConfig.Enabled`
- L1 cost computation is now split between:
  - The L1 gas total, cached in `Transaction` for performance (repeatedly used in tx pool and state transition etc.)
  - The L1 fee parameters, cached in a closure, which is part of the `BlockContext`, just like other state related tx context
- The `ApplyTransaction` now behaves the same as the block processing, fixing the caching bug. The cache is not filled when deposits are processed, so it'll use the latest l1 info data (introduced with a deposit at start of block)
- Fix circular dependency issue with contract addresses.
- Fix tx pool: transactions now get filtered based on L1 cost, both when adding, and when promoting/demoting txs (when fees are checked too).
- Fix RPC: L1 cost is now applied everywhere, by initializing the `BlockContext` everywhere where `AsMessage` was being called.
  - Historical state production with tx replay
  - Tracing RPC

Works with op-e2e tests on monorepo `develop` branch

### CODE REVIEW

File: core/rollup_l1_cost.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+// NewL1CostFunc returns a function used for calculating L1 fee cost.
+// This depends on the oracles because gas costs can change over time.
+// It returns nil if there is no applicable cost function.
+func NewL1CostFunc(config *params.ChainConfig, statedb vm.StateDB) vm.L1CostFunc {
+	cacheBlockNum := ^uint64(0)
+	var l1BaseFee, overhead, scalar, decimals, divisor *big.Int
+	return func(blockNum uint64, msg vm.RollupMessage) *big.Int {
+		rollupDataGas := msg.RollupDataGas() // Only fake txs for RPC view-calls are 0.
+		if config.Optimism == nil || msg.Nonce() == types.DepositTxType || rollupDataGas == 0 {
```
Comment (trianglesphere) [inline core/rollup_l1_cost.go:50]: nonce and `types.DepositTxType` are not logically compatible. I think it's the DepositNonceValue here

File: core/rollup_l1_cost.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+			return nil
+		}
+		if blockNum != cacheBlockNum {
```
Comment (trianglesphere) [inline core/rollup_l1_cost.go:53]: Relies on the invariant that deposits are first & no deposits later in the block. Nice way of pulling in the values when needed.

**Review by trianglesphere — COMMENTED** [review]
I gave parts of this a light skim. The context functions seems like it makes sense. I bet I can rework the unmetering PR to not use the msg options interface.

File: core/rollup_l1_cost.go (hunk 1 above)
Comment (protolambda) [reply core/rollup_l1_cost.go]: Nice catch! Checked all usages of `DepositTxType` and `DepositsNonce` to make sure I didn't mix them up elsewhere. Fixed :+1:

File: core/rollup_l1_cost.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+var (
+	L1BaseFeeSlot = common.BigToHash(big.NewInt(1))
+	OverheadSlot  = common.BigToHash(big.NewInt(3))
+	ScalarSlot    = common.BigToHash(big.NewInt(4))
+	DecimalsSlot  = common.BigToHash(big.NewInt(5))
+)
+
+var (
+	OVM_GasPriceOracleAddr = common.HexToAddress("0x420000000000000000000000000000000000000F")
```
Comment (tuxcanfly) [inline core/rollup_l1_cost.go:38]: Might want to remove the `OVM_` prefix.

File: params/config.go [hunk 4]
```diff
@@ -383,7 +383,6 @@ func (c *CliqueConfig) String() string {
 
 // OptimismConfig is the optimism config.
 type OptimismConfig struct {
-	Enabled          bool           `json:"enabled"`
 	BaseFeeRecipient common.Address `json:"baseFeeRecipient"`
```
[the remark this answers is not in the document]
Comment (protolambda) [reply params/config.go:386]: Yes, thought about that too, but don't want to scope-creep this PR too much.

File: core/rollup_l1_cost.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	L1BlockAddr            = common.HexToAddress("0x4200000000000000000000000000000000000015")
+)
+
+// NewL1CostFunc returns a function used for calculating L1 fee cost.
+// This depends on the oracles because gas costs can change over time.
```
[the remark this answers is not in the document]
Comment (protolambda) [reply core/rollup_l1_cost.go:43]: We have L1 cost checking in op-e2e, we try to do testing there to keep the diff simple and small.

File: core/types/transaction.go [hunk 6]
```diff
@@ -330,6 +333,34 @@ func (tx *Transaction) Cost() *big.Int {
...
+func (tx *Transaction) RollupDataGas() uint64 {
+	if tx.Type() == DepositTxType {
+		return 0
+	}
+	if v := tx.rollupGas.Load(); v != nil {
+		return v.(uint64)
+	}
+	data, err := tx.MarshalBinary()
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (protolambda) [reply core/types/transaction.go]: Removed the panic, replaced with a log. Added a comment the error is silent. If tx encoding fails (it should never), then it won't be batch-submitted or readable anyway.

File: core/rollup_l1_cost.go (hunk 3 above)
Comment (protolambda) [reply core/rollup_l1_cost.go:38]: The prefix is being removed from the contracts, but the monorepo Go code still has the prefix, so I'm going to keep it for now.

File: core/types/transaction.go [hunk 7]
```diff
@@ -330,6 +334,34 @@ func (tx *Transaction) Cost() *big.Int {
 	return total
 }
 
+// RollupDataGas is the amount of gas it takes to confirm the tx on L1 as a rollup
+func (tx *Transaction) RollupDataGas() uint64 {
```
Comment (tynes) [inline core/types/transaction.go:338]: This function will need to be hardforkable eventually

Comment (protolambda) [reply core/types/transaction.go:338]: Worst-case we deprecate the function, and call a different function. The l1 cost computation has access to both the chain config and current block number, so there it's easy to hardfork: https://github.com/ethereum-optimism/reference-optimistic-geth/blob/d9da8f7e57d9770fed3b79f537152985e28dc8cd/core/rollup_l1_cost.go#L49

File: core/tx_pool.go [hunk 8]
```diff
@@ -1539,8 +1559,15 @@ func (pool *TxPool) demoteUnexecutables() {
 			pool.all.Remove(hash)
 			log.Trace("Removed old pending transaction", "hash", hash)
 		}
+		balance := pool.currentState.GetBalance(addr)
+		if !list.Empty() {
+			// Reduce the cost-cap by L1 rollup cost of the first tx if necessary. Other txs will get filtered out afterwards.
```
Comment (trianglesphere) [inline core/tx_pool.go:1564]: I think this works, but it's a bit weird b/c there's no hard cap.

If there was a hard cap we could do something like enforce that the balance was enough to pay the cap & then place it into the tx priced list. Right now without the cap, it's not really clear when the tx would/could become valid again if it had low but present funds.

**Review by trianglesphere — APPROVED** [review]
The only place I'm a little iffy on is the tx pool. I think it does work for now, but the semantics around the L1 Cost make it hard to integrate it into the tx pool.

**Comment by protolambda** [comment]
Seems like CI is flaky, `TestFakedSyncProgress66Full` runs fine for me locally and seems unrelated. Re-running it now

**Pull request merged**